src/NuGet.Core/NuGet.Common/MsBuildStringUtility.cs

```diff
@@ -85,19 +85,32 @@ public static bool IsTrueOrEmpty(string? value)
         }
 
         /// <summary>
-        /// Splits and parses a ; or , delimited list of log codes.
-        /// Ignores codes that are unknown.
+        /// Parses the specified string as a comma or semicolon delimited list of NuGet log codes and ignores unknown codes.
         /// </summary>
-        public static IEnumerable<NuGetLogCode> GetNuGetLogCodes(string s)
+        /// <param name="s">A comma or semicolon delimited list of NuGet log codes.</param>
+        /// <returns>An <see cref="IList{T}" /> containing the <see cref="NuGetLogCode" /> values that were successfully parsed from the specified string.</returns>
+        public static IList<NuGetLogCode> GetNuGetLogCodes(string s)
         {
-            foreach (var item in MSBuildStringUtility.Split(s, ';', ','))
+            // The Split() method already checks for an empty string and returns Array.Empty<string>().
+            string[] split = MSBuildStringUtility.Split(s, ';', ',');
```

Comment:
I wouldn't be surprised if the allocs in `MSBuildStringUtility.Split()` show up here. Here's how it's implemented:

        public static string[] Split(string s, params char[] chars)
        {
            if (!string.IsNullOrEmpty(s))
            {
                // Split on ; and trim all entries
                // After trimming remove any entries that are now empty due to trim.
                return s.Split(chars, StringSplitOptions.RemoveEmptyEntries)
                    .Select(entry => entry.Trim())
                    .Where(entry => entry.Length != 0)
                    .ToArray();
            }
            return Array.Empty<string>();
        }

Each call that actually does any work is going to allocate:

- A `char[]` because of params
- a `string[]` returned from `Split()`
- an `int[]` the size of the input `string` allocated internally by `Split()`
- An alloc for each `string` in the `string[]`
- Boxing of the `char[]` enumerator from `Select()`
- Potentially new strings from the call to `Trim()`
- a new `string[]` from the call to `ToArray()`

We've seen lots of allocations arising from code that splits strings, and like the code below, the string may not even be needed (only needed if it starts with `"NU"`). We ended up writing a helper function that splits the string without any allocations unless you actually need to materialize the string. Instead, you get a struct enumerator of the results of the split operation as `ReadOnlySpan<char>`. If this is allocating a sufficient amount (here or cumulatively in all `Split()` calls), I'd consider porting the split helper.

Reply:
https://github.com/NuGet/NuGet.Client/blob/e520f969f09411cdbf024349cbae3a1f60c4c73c/src/NuGet.Core/NuGet.Common/MsBuildStringUtility.cs#L25-L37